File: pdu/python/geometry_msgs/pdu_conv_TwistWithCovariance.py

import struct
from .pdu_pytype_TwistWithCovariance import TwistWithCovariance
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..geometry_msgs.pdu_conv_Twist import *



def pdu_to_py_TwistWithCovariance(binary_data: bytearray) -> TwistWithCovariance:
    py_obj = TwistWithCovariance()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_TwistWithCovariance(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_TwistWithCovariance(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: TwistWithCovariance, base_off: int):
    # array_type: single 
    # data_type: struct 
    # member_name: twist 
    # type_name: Twist 
    # offset: 0 size: 48 
    # array_len: 1

    tmp_py_obj = Twist()
    binary_read_recursive_Twist(meta, binary_data, tmp_py_obj, base_off + 0)
    py_obj.twist = tmp_py_obj
    
    # array_type: array 
    # data_type: primitive 
    # member_name: covariance 
    # type_name: float64 
    # offset: 48 size: 288 
    # array_len: 36

    
    array_value = binary_io.readBinary(binary_data, base_off + 48, 288)
    py_obj.covariance = binary_io.binToArrayValues("float64", array_value)
    
    return py_obj


def py_to_pdu_TwistWithCovariance(py_obj: TwistWithCovariance) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_TwistWithCovariance(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_TwistWithCovariance(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: TwistWithCovariance):
    # array_type: single 
    # data_type: struct 
    # member_name: twist 
    # type_name: Twist 
    # offset: 0 size: 48 
    # array_len: 1
    type = "Twist"
    off = 0

    binary_write_recursive_Twist(parent_off + off, bw_container, allocator, py_obj.twist)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: covariance 
    # type_name: float64 
    # offset: 48 size: 288 
    # array_len: 36
    type = "float64"
    off = 48

    
    elm_size =  288 
    array_size = int(8.0)
    one_elm_size = int(elm_size / array_size)
    binary = binary_io.typeTobin_array(type, py_obj.covariance, one_elm_size)
    allocator.add(binary, expected_offset=(parent_off + off))
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_TwistWithCovariance <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_TwistWithCovariance(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = TwistWithCovariance.from_json(json_str)
        
        binary_data = py_to_pdu_TwistWithCovariance(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/geometry_msgs/pdu_conv_TwistWithCovariance.js
import * as PduUtils from '../pdu_utils.js';
import { TwistWithCovariance } from './pdu_jstype_TwistWithCovariance.js';
import { Twist } from './pdu_jstype_Twist.js';
import { binary_read_recursive_Twist, binary_write_recursive_Twist } from './pdu_conv_Twist.js';


/**
 * Deserializes a binary PDU into a TwistWithCovariance object.
 * @param {ArrayBuffer} binary_data
 * @returns { TwistWithCovariance }
 */
export function pduToJs_TwistWithCovariance(binary_data) {
    const js_obj = new TwistWithCovariance();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_TwistWithCovariance(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_TwistWithCovariance(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: twist, type: Twist (struct)

    {
        const tmp_obj = new Twist();
        binary_read_recursive_Twist(meta, binary_data, tmp_obj, base_off + 0);
        js_obj.twist = tmp_obj;
    }
    
    // member: covariance, type: float64 (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 48, 288);
        js_obj.covariance = PduUtils.binToArrayValues("float64", array_bin, 36);
    }
    
    return js_obj;
}

/**
 * Serializes a TwistWithCovariance object into a binary PDU.
 * @param { TwistWithCovariance } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_TwistWithCovariance(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_TwistWithCovariance(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_TwistWithCovariance(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: twist, type: Twist (struct)

    {
        binary_write_recursive_Twist(parent_off + 0, bw_container, allocator, js_obj.twist);
    }
    
    // member: covariance, type: float64 (primitive)

    
    {
        const buffer = PduUtils.typesToBin("float64", js_obj.covariance);
        allocator.add(buffer, parent_off + 48);
    }
    
}
